[SR-69][Sema] Improved quality of diagnoses messages

gregomni opened the pull request:
In the specific case of sr-69, and in a bunch of other code where errors arise involving generic function application, better type constraint failure diagnoses are being masked by the overly conservative implementation in evaluateCloseness(). If the actual arg types didn’t exactly match the parameter types, we’d always diagnose a non-specific arguments-don’t-match error instead of allowing discovery of better errors from the constraint system.

This commit adds more cases where evaluateCloseness will return CC_ExactMatch, specifically in application of functions with one or more arguments of a single archetype, like `func min<T: Comparable>(T, T) -> T`. It verifies that the actual argument type isSubstitutableFor() the archetype, and that all such arguments are of the same type. Anything more complicated than that still has the previous behavior of not matching at all. 

I think the final answer here ought to be to make a constraint system with type variables for any archetypes, add appropriate constraints to the actual args and then see if the system can solve all the argument constraints at once. That’s because the next most complicated set of things to handle in the stdlib are things like `func -<T: Strideable>(lhs: T, rhs: T.Stride)` where generic argument types depend on each other. I tried attacking that, but it was too big of a bite for me to manage all at once. But there are FIXME’s here to try that again at some point.

New tests for SR-69 are at the end of deduction.swift, and the rest of the test changes are generally improved deduced diagnoses. I think the changed diagnoses in materializable_restrictions.swift is the only one which is worse instead of better, and that’s just because the previous general message mentioned `inout` basically accidentally. Opportunity for further improvement (a new diagnosis maybe) there. 

Validation tests run and passed.

lattner commented:
Wow, this is fantastic, thank you for working on this.  Here are some nit-picky comments:

- Instead of:
 paramType->getKind() == TypeKind::Archetype

Please use this, it is a bit more idiomatic (generalizes to dyn_cast etc)
  paramType->is<ArchetypeType>()

- It looks like line 1240 got 4 spaces of indentation instead of 2.

In this:

+         // Since we've 'bound' singleArchetype and it is now potentially wrong,
 +              // this is then the second mismatch -- give up right away.
 +              return { CC_ArgumentMismatch, {}};
 
we're losing the resolution of "CC_ArgumentNearMismatch".  

- As a followup commit, it probably makes sense to move evaluateCloseness to being a method on CalleeCandidateInfo.  Earlier in the compiler this wasn't feasible, but it looks like it is now and will allow avoid having to pass in CS & argsHaveTrailingClosure.

gregomni commented:
@lattner Okay, cleaned up ->is() and the spacing. (and squashed.) 

There is no actual behavioral change for near mismatches, since argumentMismatchIsNearMiss() also does isEqual() comparisons to the ArchetypeType paramTypes and so would always return false for these args.

gregomni commented:
Oh, and thank you for the wonderful FIXME comments! (Here specifically, and throughout the project generally.) I spent awhile tracing things down to here and when I found them had the somewhat mixed feeling of "oh, if only I'd not wasted my time in the debugger and had seen this earlier" combined with "well, at least I really know this is the right place to fix things".

gregomni commented:
Oops! The ->is<ArchetypeType>() change caused a couple additional tests to be effected, since it does desugaring of the type before testing so the generic checks got hit a little bit more often. Changed a couple more expected-error results, committed, squashed.

gregomni commented:
It occurred to me that I was being dumb about the near mismatches thing and hadn't got what you almost surely meant, which is that you can compare the mismatching arg for near miss'ness to the type that you bound the archetype to. I was concentrating on exact matches and not worrying about improving other closeness results. Did that here as a separate commit (and threw in making it into a member) to check that that's what you intended to tell me. 

And as long as this is now trying to be smarter about partial matching of generics as well as exact matching, changed the logic a bit so that if an arg type isn't substitutable it doesn't 'bind' it at all. So (with non-Comparable struct A) `min(A(), 3)` becomes a one arg mismatch just like `min(3, A())`.

If this looks good to you there will be more test result changes to make before merging.

gregomni commented:
Producing single argument mismatches involving generics causes some gross looking error messages if the generic mismatches get put into the same closeness bucket as non-generic mismatches.

E.g. `var v71 = true + 1.0` used to produce `error: cannot convert value of type 'Bool' to expected argument type 'Double’`, but would now end up with `binary operator '+' cannot be applied to operands of type 'Bool' and 'Double’` `overloads for '+' exist with these partially matching parameter lists: (Double, Double), (T, T.Stride), (T.Stride, T)`.

Resolve this by adding CC_OneGenericArgumentNearMismatch and CC_OneGenericArgumentMismatch, that are similar but ever so slightly not as close as a mismatch involving non-generic functions. This gets back the original error message in cases like the above, because there is only one declaration of `+` which partially matches and is non-generic, and the generic partial matches are now farther away.

But now single arg mismatches and nearness work for single-archetype generic functions, as in the additions to the SR-69 test at the end of deduction.swift.

lattner commented:
Sounds good, thanks for working on this!
